04-Protocol-Security.md

```diff
@@ -0,0 +1,407 @@
+6. calls `MixKey(ECDH(s, re))`
+7. appends `EncryptAndHash(SIGNATURE_NOISE_MESSAGE)` to the buffer
+8. submits the buffer for sending to the initiator
+9. return pair of CipherState objects, the first for encrypting transport messages from initiator to responder, and the second for messages in the other direction:
+   1. sets `temp_k1, temp_k2 = HKDF(ck, zerolen, 2)`
+   2. creates two new CipherState objects `c1` and `c2`
+   3. calls `c1.InitializeKey(temp_k1)` and `c2.InitializeKey(temp_k2)`
+   4. returns the pair `(c1, c2)`
+
+```
+Message format of NX-handshake part 2 
++-------------------------+--------------------------------------------------------------------------------------------+
+| PUBKEY                  |  Responder's plaintext ephemeral public key                                                |
++-------------------------+--------------------------------------------------------------------------------------------+
+| PUBKEY                  |  Responder's encrypted static public key                                                   |
++-------------------------+--------------------------------------------------------------------------------------------+
+| MAC                     |  Message authentication code for responder's static public key                             |
++-------------------------+--------------------------------------------------------------------------------------------+
+| SIGNATURE_NOISE_MESSAGE |  Signed message containing Responder's static key. Signature is issued by authority that   |
+|                         |  is generally known to operate the server acting as the noise responder                    |
++-------------------------+--------------------------------------------------------------------------------------------+
+| MAC                     |  Message authentication code for SIGNATURE_NOISE_MESSAGE                                   |
++-------------------------+--------------------------------------------------------------------------------------------+
+
+Message length: 170 bytes
+```
+
+#### 4.5.2.2 Initiator
+1. receives NX-handshake part 2 message
+2. interprets first 32 bytes as `re.public_key`
+3. calls `MixHash(re.public_key)`
+4. calls `MixKey(ECDH(e, re))`
+5. decrypts next 48 bytes with `DecryptAndHash()` and stores the results as `rs.public_key` which is **server's static public key** (note that 32 bytes is the public key and 16 bytes is MAC)
+6. calls `MixKey(ECDH(e, rs)`
+7. decrypts next 90 bytes with `DecryptAndHash()` and deserialize plaintext into `SIGNATURE_NOISE_MESSAGE` (74 bytes data + 16 bytes MAC)
+9. return pair of CipherState objects, the first for encrypting transport messages from initiator to responder, and the second for messages in the other direction:
+   1. sets `temp_k1, temp_k2 = HKDF(ck, zerolen, 2)`
+   2. creates two new CipherState objects `c1` and `c2`
+   3. calls `c1.InitializeKey(temp_k1)` and `c2.InitializeKey(temp_k2)`
```

Comment:
CAREFUL! It seems that initiator and responder have the same encryption key enc_k and the same
decryption key dec_k. So, each party would use dec_k for decrypting a message encrypted with enc_k,
a different key. I didn't look deeply at ChaCha20-Poly1305, but in a symmetric key algorithms,
a message encrypted with a key k can be decrypted only with the key k. Consider switching the 
keys for one partcipant; for example, at line 263, for the initiator:
c1.InitializeKey(temp_k2) and c2.InitializeKey(temp_k1)

Reply:
> All messages that initiator sends are encrypted with temp_k1. All messages that responder receives are decrypted with temp_k1.
> And correspondingly, all messages that initiator receives are decrypted with temp_k2 and all messages that responder sends are encrypted with temp_k2.

Yes that's what I meant. I think that you should make it a bit more evident.